### PR TITLE
Make `in` a first-class storage class, not a `const [scope]` alias

### PR DESCRIPTION
**Pull request by Geod24**
Splitted off #11000 . I believe this changeset is useful on its own, but #11000 might provide an additional reasoning for why we need it.

Companion Phobos PR: dlang/phobos#7570
Spec PR / changelog pending more discussions, including #11000 .

Commit message follows:
```
For a long time, the 'in' storage class was only a second class citizen,
merely an alias for 'const', which is a type constructor.
While it was also documented for a long time as being 'scope',
this was removed when 'scope' actually started to have an effect
(when DIP1000 started to be implemented).
Currently, a switch (-preview=in) allows to get back this 'scope'.

However, the 'in' storage class does not really exists,
it gets lowered to 'const [scope]' at an early stage by the compiler,
which means that we expose what is essentially an implementation
detail to the user.

This led to a variety of problems for both developers and users.
The most obvious one was that functions involving `in` would show
up simply as `const` (or `const scope`), both in generated header
and error messagges, hindering the ability to do header generation
with `-preview=in` as the same value for this switch was needed
for both producer and consumer.
Another issue was that the result of `__traits(getParameterStorageClass)`
was inaccurate, giving either an empty tuple or `scope`.

For compiler developers, the `in` storage class simply couldn't be relied on,
as it was replaced by `const [scope]` rather early in the semantic phase
(in TypeFunction's semantic), which lead to some dead code (e.g. in dmangle).

After this change, the class will show up properly in error message,
and in `getParameterStorageClass`, which can be a surprising change for users.
However, it is not expected that code can break as a result,
unless they break as a result of using the `-preview=in` switch.
```

### CODE REVIEW

**Comment by Geod24** [comment]
One question I have is, should `in` have its own mangling ? Currently it does not, so applications compiled with different values for `-transition=in` will not link together. It also means that demangling-based name printing (e.g. stack traces) will still be incorrect. And if it does, what value should it take ? CC @ibuclaw @rainers .

**Comment by rainers** [comment]
My proposal (https://github.com/dlang/dmd/pull/10526) was to add `in` to the mangling. It seems your version is a bit more advanced. IIRC I had some issues building phobos unittest, did you run into them, too?

**Comment by Geod24** [comment]
> My proposal (#10526) 

Thanks, I have integrated that part now.

> IIRC I had some issues building phobos unittest, did you run into them, too?

I mostly had issues with covariance (you can see that's a good chunk of the tests).
Note that I built this as part of #11000 first, and tested that PR with my project (depending on Vibe.d, D-YAML and a bunch of other things) to be confident it worked. Once it was working, carving this out was trivial.

I think the main difference with your approach is that I don't try to infer `in` at all, and that except for the switch rename, none of this PR is about DIP1000. It's simply about pushing the lowering of `in` to a later stage / keeping the STC around. The beauty of providing a mangling for `in` is that we completely sidestep the issue of compiling Phobos with `-preview=in` in its current form (obviously you forgo the safety checks for non-template, but that's not really an issue).

I just ran the unittests with the companion Phobos PR mentioned in the op and it passed (OSX).

**Comment by rainers** [comment]
> I mostly had issues with covariance (you can see that's a good chunk of the tests).

Maybe the issues I noticed where related to that.

> I think the main difference with your approach is that I don't try to infer `in` at all

"Inferring scope" (I guess that's what you meant) was probably not the correct term. It referred to the fact that it doesn't appear in the mangling, because it uses `STC.scopeinferred`. I think you might want to add it, too, to avoid linking errors with libraries compiled differently.

**Comment by rainers** [comment]
With the mangling added, we might want to resurrect https://github.com/dlang/druntime/pull/2847

**Comment by Geod24** [comment]
> It referred to the fact that it doesn't appear in the mangling [...]

Neither does it here. It's either `in` or `scope` (`scope` and `in` are considered redundant by the parser).
The mangling / header / error messages are derived from the `Parameter` class, which keeps `STC.in_` through its lifetime.
On the other hand, the checks on `scope` are done on the `VarDeclaration`, which will get `STC.scope_`.

**Comment by rainers** [comment]
> The mangling / header / error messages are derived from the `Parameter` class, which keeps `STC.in_` through its lifetime.
> On the other hand, the checks on `scope` are done on the `VarDeclaration`, which will get `STC.scope_`.

Ah, ok. Thanks for explanation.

**Review by rainers — APPROVED** [review]

**Comment by Geod24** [comment]
The Phobos fix has been merged, this should now go green on the auto-tester.
Removed some of the tests that were tied to `-preview=in`, in particular the covariance ones. Both the test-suite / druntime covers it already, and they are still in the other PR as well.
Vibe-core needs a fix because it is testing the exact output of a `stringof` call in a `static assert`: https://github.com/vibe-d/vibe-core/pull/213

**Comment by Geod24** [comment]
`vibe-core` has been fixed, this is now green.

**Review by thewilsonator — APPROVED** [review]
Looks fine, no tests for mangles though.

**Comment by Geod24** [comment]
Added a mangling test (piggybacking on the fail_compilation test, but it works).

**Comment by Geod24** [comment]
Actually, I also added a changelog to explain what happens (and one of the possible shortcomings).

File: src/dmd/dmangle.d [hunk 1]
```diff
@@ -79,7 +79,7 @@ private immutable char[TMAX] mangleChar =
     Tfunction    : 'F', // D function
     Tsarray      : 'G',
     Taarray      : 'H',
-    Tident       : 'I',
```
Comment (UplinkCoder) [inline src/dmd/dmangle.d:82]: I am wondering why this change does not seem to break tests.
Where was Tident used? in templates?

Comment (Geod24) [reply src/dmd/dmangle.d:82]: It isn't used (https://github.com/dlang/druntime/pull/2847#discussion_r341837820)

**Review by AndrejMitrovic — APPROVED** [review]

**Comment by AndrejMitrovic** [comment]
Seems fairly uncontroversial.
